File: jtaf-web/src/main/java/ch/jtaf/report/EventsRanking.java
package ch.jtaf.report;

import ch.jtaf.to.AthleteWithEventTO;
import ch.jtaf.data.EventsRankingData;
import ch.jtaf.data.EventsRankingEventData;
import com.itextpdf.text.Document;
import com.itextpdf.text.DocumentException;
import com.itextpdf.text.PageSize;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import org.jboss.logging.Logger;

public class EventsRanking extends Ranking {

    private Document document;
    private PdfWriter pdfWriter;
    private final EventsRankingData ranking;

    public EventsRanking(EventsRankingData ranking) {
        this.ranking = ranking;
    }

    public byte[] create() {
        try {
            byte[] ba;
            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
                float border = cmToPixel(1.5f);
                document = new Document(PageSize.A4, border, border, border, border);
                pdfWriter = PdfWriter.getInstance(document, baos);
                pdfWriter.setPageEvent(new HeaderFooter(
                        "Event Ranking", ranking.getCompetition().getName(),
                        sdf.format(ranking.getCompetition().getCompetitionDate())));
                document.open();
                createRanking();
                document.close();
                pdfWriter.flush();
                ba = baos.toByteArray();
            }

            return ba;
        } catch (DocumentException | IOException e) {
            Logger.getLogger(EventsRanking.class).error(e.getMessage(), e);
            return new byte[0];
        }
    }

    private void createRanking() throws DocumentException {
        for (EventsRankingEventData event : ranking.getEvents()) {
            PdfPTable table = new PdfPTable(new float[]{2f, 10f, 10f, 2f, 2f, 5f, 5f});
            table.setWidthPercentage(100);
            table.setSpacingBefore(cmToPixel(1f));

            createEventTitle(table, event);

            int position = 1;
            for (AthleteWithEventTO athlete : event.getAthletes()) {
                createAthleteRow(table, position, athlete);
                position++;
            }
            document.add(table);
        }
    }

    private void createEventTitle(PdfPTable table, EventsRankingEventData event) {
        addCategoryTitleCellWithColspan(table, event.getEvent().getName(), 2);
        addCategoryTitleCellWithColspan(table, event.getEvent().getLongName(), 3);
        addCategoryTitleCellWithColspan(table, event.getEvent().getGender(), 2);


        addCategoryTitleCellWithColspan(table, " ", 7);
    }

    private void createAthleteRow(PdfPTable table, int position, AthleteWithEventTO athlete) throws DocumentException {
        addCell(table, position + ".");
        addCell(table, athlete.getAthlete().getLastName());
        addCell(table, athlete.getAthlete().getFirstName());
        addCell(table, athlete.getAthlete().getYear() + "");
        addCell(table, athlete.getAthlete().getCategory().getAbbreviation() + "");
        addCell(table, athlete.getAthlete().getClub() == null ? "" : athlete.getAthlete().getClub().getAbbreviation());
        addCellAlignRight(table, "" + athlete.getResult().getResult());
    }

}

File: src/main/kotlin/ch/jtaf/control/reporting/report/EventsRanking.kt
package ch.jtaf.control.reporting.report

import ch.jtaf.control.reporting.data.EventsRankingData
import ch.jtaf.control.reporting.data.EventsRankingEventData
import ch.jtaf.entity.Result
import com.itextpdf.text.Document
import com.itextpdf.text.PageSize
import com.itextpdf.text.pdf.PdfPTable
import com.itextpdf.text.pdf.PdfWriter
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.util.*

class EventsRanking(private val ranking: EventsRankingData, locale: Locale) : Ranking(locale) {

    private val logger = LoggerFactory.getLogger(EventsRanking::class.java)

    private var document: Document? = null

    fun create(): ByteArray {
        try {
            ByteArrayOutputStream().use { baos ->
                val border = cmToPixel(1.5f)
                document = Document(PageSize.A4, border, border, border, border)
                val pdfWriter = PdfWriter.getInstance(document!!, baos)
                pdfWriter.pageEvent = HeaderFooter("Event Ranking", ranking.competition.name, sdf.format(ranking.competition.competitionDate))
                document!!.open()
                createRanking()
                document!!.close()
                pdfWriter.flush()
                return baos.toByteArray()
            }
        } catch (e: Exception) {
            logger.error(e.message, e)
            return ByteArray(0)
        }
    }

    private fun createRanking() {
        ranking.events.forEach {
            val table = PdfPTable(floatArrayOf(2f, 10f, 10f, 2f, 2f, 5f, 5f))
            table.widthPercentage = 100f
            table.spacingBefore = cmToPixel(1f)
            table.keepTogether = true

            createEventTitle(table, it)

            var position = 1
            it.athletes.forEach {
                createAthleteRow(table, position, it)
                position++
            }
            document!!.add(table)
        }
    }

    private fun createEventTitle(table: PdfPTable, event: EventsRankingEventData) {
        addCategoryTitleCellWithColspan(table, event.event.name + " / " + event.event.gender, 7)

        addCategoryTitleCellWithColspan(table, " ", 7)
    }

    private fun createAthleteRow(table: PdfPTable, position: Int, result: Result) {
        addCell(table, position.toString() + ".")
        addCell(table, result.athlete!!.lastName)
        addCell(table, result.athlete!!.firstName)
        addCell(table, result.athlete!!.yearOfBirth.toString())
        addCell(table, result.category!!.abbreviation)
        addCell(table, if (result.athlete!!.club == null) "" else result.athlete!!.club?.abbreviation!!)
        addCellAlignRight(table, result.result)
    }

}
